### PR TITLE
Set form options for form-based request types

### PR DESCRIPTION
**Pull request by captainsafia**
Part of https://github.com/dotnet/aspnetcore/issues/50008.

### CODE REVIEW

**Comment by BrennanConroy** [comment]
@sebastienros Does the bot not write a comment with results anymore?

@captainsafia The "platform" benchmarks don't use Routing so won't show any difference here. The "mvcjsoninput2k" benchmark might show a small difference, ~615k on main -> ~625k a couple days ago; if I'm looking at the correct benchmark on the dashboard.

**Comment by captainsafia** [comment]
> @captainsafia The "platform" benchmarks don't use Routing so won't show any difference here. The "mvcjsoninput2k" benchmark might show a small difference, ~615k on main -> ~625k a couple days ago; if I'm looking at the correct benchmark on the dashboard.

Yes, @sebastienros and I were discussing last night that there's currently no way for us to test PRs with changes in the routing layer. Using the `mvcjsoninput2k` build sounds like a good idea, but I still think it's worthwhile for us to get an "endpoint" component configured in the benchmark bot so we can profile changes to Microsoft.AspNetCore.Routing in PRs.

It's apparently straightforward to add support for a new component type so I'll try but in the meantime I'll also run them locally.

File: src/Shared/HttpExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.Net.Http.Headers;
+
+internal static class HttpExtensions
+{
+    internal static bool IsValidHttpMethodForForm(string method) =>
+        HttpMethods.IsPost(method) || HttpMethods.IsPut(method) || HttpMethods.IsPatch(method);
+
+    internal static bool IsValidContentTypeForForm(string? contentType) =>
+        MediaTypeHeaderValue.TryParse(contentType, out var mediaType) &&
```
Comment (davidfowl) [inline src/Shared/HttpExtensions.cs:12]: This allocates, no?

File: src/Http/Routing/src/EndpointRoutingMiddleware.cs [hunk 2]
```diff
@@ -139,8 +139,14 @@ private Task SetRoutingAndContinue(HttpContext httpContext)
             // We do this during endpoint routing to ensure that successive middlewares in the pipeline
             // can access the feature with the correct value.
             SetMaxRequestBodySize(httpContext);
-            // Map IFormOptionsMetadata to IFormFeature if present on the endpoint.
-            SetFormOptions(httpContext, endpoint);
+            // Map IFormOptionsMetadata to IFormFeature if present on the endpoint if the
+            // request being processed is a form request. Note: we do a manual check for the
+            // form content-types here to avoid prematurely accessing the form feature.
+            if (HttpExtensions.IsValidHttpMethodForForm(httpContext.Request.Method) &&
+                HttpExtensions.IsValidContentTypeForForm(httpContext.Request.ContentType))
```
[the remark this answers is not in the document]
Comment (campersau) [reply src/Http/Routing/src/EndpointRoutingMiddleware.cs:146]: Currently it is possible to manually set the `Form` and `HasFormContentType` would return `true` indenpendent of the actual header values.

https://github.com/dotnet/aspnetcore/blob/d9a1cf0a0315c6da156dc281469f060f5bdc7b49/src/Http/Http/src/Features/FormFeature.cs#L72-L85

[the remark this answers is not in the document]
Comment (davidfowl) [reply src/Http/Routing/src/EndpointRoutingMiddleware.cs:146]: > Currently it is possible to manually set the Form and HasFormContentType would return true indenpendent of the actual header values.

I don't consider that a problem.

File: src/Shared/HttpExtensions.cs (hunk 1 above)
Comment (captainsafia) [reply src/Shared/HttpExtensions.cs:12]: Mmmm, yeah, and we would be paying the cost here for most requests (`POST`, `PUT`). I think I can cook up something better here.

**Comment by captainsafia** [comment]
Baseline:

| application             |                             |
| ----------------------- | --------------------------- |
| Max CPU Usage (%)       | 99                          |
| Max Cores usage (%)     | 2,759                       |
| Max Working Set (MB)    | 520                         |
| Max Private Memory (MB) | 1,277                       |
| Build Time (ms)         | 4,159                       |
| Start Time (ms)         | 168                         |
| Published Size (KB)     | 116,042                     |
| Symbols Size (KB)       | 52                          |
| .NET Core SDK Version   | 8.0.100-rc.1.23411.1        |
| ASP.NET Core Version    | 8.0.0-rc.1.23411.1+d9a1cf0  |
| .NET Runtime Version    | 8.0.0-rc.1.23410.15+786b987 |


| load                    |                |
| ----------------------- | -------------- |
| Max CPU Usage (%)       | 41             |
| Max Cores usage (%)     | 1,136          |
| Max Working Set (MB)    | 48             |
| Max Private Memory (MB) | 358            |
| Build Time (ms)         | 3,062          |
| Start Time (ms)         | 0              |
| Published Size (KB)     | 73,203         |
| Symbols Size (KB)       | 0              |
| .NET Core SDK Version   | 7.0.400        |
| ASP.NET Core Version    | 7.0.10+5a4c82e |
| .NET Runtime Version    | 7.0.10+a6dbb80 |
| First Request (ms)      | 101            |
| Requests/sec            | 3,808,360      |
| Requests                | 57,502,947     |
| Mean latency (ms)       | 0.71           |
| Max latency (ms)        | 43.67          |
| Bad responses           | 0              |
| Socket errors           | 0              |
| Read throughput (MB/s)  | 592.01         |
| Latency 50th (ms)       | 0.61           |
| Latency 75th (ms)       | 0.90           |
| Latency 90th (ms)       | 1.21           |
| Latency 99th (ms)       | 0.00           |


https://github.com/dotnet/aspnetcore/pull/50016/commits/2e6e1d5462d6d1182a44b3fece0d2be4f5a5ef44:


| application             |                             |
| ----------------------- | --------------------------- |
| Max CPU Usage (%)       | 99                          |
| Max Cores usage (%)     | 2,780                       |
| Max Working Set (MB)    | 521                         |
| Max Private Memory (MB) | 1,277                       |
| Build Time (ms)         | 6,232                       |
| Start Time (ms)         | 169                         |
| Published Size (KB)     | 116,042                     |
| Symbols Size (KB)       | 52                          |
| .NET Core SDK Version   | 8.0.100-rc.1.23411.1        |
| ASP.NET Core Version    | 8.0.0-rc.1.23411.1+d9a1cf0  |
| .NET Runtime Version    | 8.0.0-rc.1.23410.15+786b987 |


| load                    |                |
| ----------------------- | -------------- |
| Max CPU Usage (%)       | 42             |
| Max Cores usage (%)     | 1,176          |
| Max Working Set (MB)    | 48             |
| Max Private Memory (MB) | 358            |
| Build Time (ms)         | 3,012          |
| Start Time (ms)         | 0              |
| Published Size (KB)     | 73,203         |
| Symbols Size (KB)       | 0              |
| .NET Core SDK Version   | 7.0.400        |
| ASP.NET Core Version    | 7.0.10+5a4c82e |
| .NET Runtime Version    | 7.0.10+a6dbb80 |
| First Request (ms)      | 102            |
| Requests/sec            | 4,020,284      |
| Requests                | 60,705,269     |
| Mean latency (ms)       | 0.70           |
| Max latency (ms)        | 33.81          |
| Bad responses           | 0              |
| Socket errors           | 0              |
| Read throughput (MB/s)  | 624.95         |
| Latency 50th (ms)       | 0.60           |
| Latency 75th (ms)       | 0.87           |
| Latency 90th (ms)       | 1.24           |
| Latency 99th (ms)       | 0.00           |

File: src/Shared/HttpExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using Microsoft.AspNetCore.Http;
+
+internal static class HttpExtensions
+{
+    internal static string UrlEncodedFormContentType = "application/x-www-form-urlencoded";
+    internal static string MultipartFormContentType = "multipart/form-data";
```
Comment (martincostello) [inline src/Shared/HttpExtensions.cs:8]: Could/should these be constants? Then you could test against the length of the shortest one lower down and avoid the 19 magic number?

**Comment by captainsafia** [comment]
Bump for reviews!

**Review by sebastienros — APPROVED** [review]
Perf numbers are back to before the regression

**Comment by davidfowl** [comment]
Not part of this PR but we need to understand the overhead of requests that POST requests that are not form and POST requests that are forms.

**Comment by captainsafia** [comment]
> Not part of this PR but we need to understand the overhead of requests that POST requests that are not form and POST requests that are forms.

I'll take a stab at adding benchmarks in-repo for this.

**Comment by sebastienros** [comment]
We should probably track the perf of simple POST requests to detect regressions in the code that is completely excluded in GETs.

**Pull request merged**